### ISSUE TITLE
4.6.2 Intermittent Crashing - Critical Issue

### ISSUE THREAD

**Issue by phateuk** [grafana/grafana#9979]
Please include this information:
- What Grafana version are you using? v4.6.2
- What datasource are you using? Graphite v0.9.15
- What OS are you running grafana on? CentOS 6
- What did you do? Nothing, service running as normal
- What was the expected result? N/A
- What happened instead? Grafana crashes unexpectedly, no warning or errors in the log file. Just a complete crash.

So.. I upgraded our Prod Server to v4.6.2 and it was fine for 24 hours, and then crashed unexpectedly. Since then it's crashed on 4 more occasions, sometimes upto 48 hours apart, or alot less - such as today, where it crashed at 3.30am, was restarted at 7:15am and then crashed again at 10:15am.

I've seen the bug report regarding v4.6.2 and the sqllite3 segmentation fault, so I also enabled debug logging as per the below config:
```ini
[log]
# Either "console", "file", "syslog". Default is console and  file
# Use space to separate multiple modes, e.g. "console file"
;mode = console, file

# Either "trace", "debug", "info", "warn", "error", "critical", default is "info"
level = debug

# For "console" mode only
[log.console]
level = debug

# log line format, valid options are text, console and json
;format = console

# For "file" mode only
[log.file]
level = debug
```

And I am seeing 'dbug' messages now in /var/log/grafana/grafana.log. However, when it crashes, there's nothing. No errors or messages or anything, it just shows the usual INFO lines for normal user usage and then stops dead in it's tracks. I've never seen Grafana crash across all previous versions until this upgrade. Short of downgrading, I really don't know what else to do..

**Comment by torkelo** [comment]
A crash without log output, strange, that has not been reported before. Can you check memory usage and syslog?

**Comment by phateuk** [comment]
Hi Torkelo.

Thanks for the quick response. Yes it is very strange. /var/log/messages contains nothing for the times of the crashes and no other syslog files are being written to around that time. Attached is a screenshot of our CollectD dashboard for the Grafana server in question. It's a bare metal machine with 32GB of ram and no unusual resource usage at the time of the crash (10:16am), system is 99% idle. The disk usage goes to virtually zero as Grafana is no longer writing it's log file. This server literally only runs Grafana and nothing else

![grafana_collectd](https://user-images.githubusercontent.com/18192894/33208890-52da0e40-d10b-11e7-991a-1422bb5e797c.PNG)

**Comment by torkelo** [comment]
hm.. maybe an issue with go 1.9 and centos 6, this is the only report of a crash without log entry of thousands of installs. Can go through a bit more of your settings and usage? Data source, number of alerts, evaluation interval, grafana server req/s ?

**Comment by ivanich** [comment]
I suppose it's the same crashes as in https://github.com/grafana/grafana/issues/9706#
We also have grafana crashes with all 4.6.x versions at Centos6.9, ~700 dashboards, sqlite as db, graphite datasources.

**Comment by torkelo** [comment]
But that crash comes with a segmentation fault stack trace in the log. Maybe a something with your setup (disk layer etc) that makes the final crash log not flush to disk so I guess it could be same issue

**Comment by phateuk** [comment]
OK, so the out of hours engineer advised me this morning that Grafana crashed no short of 20 times on Sunday, which is odd because it was hardly being used. We had to write an auto-restart script to keep bouncing it every time it falls over.

@torkelo With regard to your questions above, is there any specific info you want me to provide? I'm not familiar re: "maybe an issue with go 1.9"?

Here are the stats from the Grafana page:
Name	Value
Total dashboards	118
Total users	78
Active users (seen last 14 days)	33
Total organizations	5
Total datasources	13
Total playlists	2
Total snapshots	24
Total dashboard tags	48
Total starred dashboards	49
Total alerts	102

Datasource is purely Graphite v0.9.15. There doesn't appear to be any issues with the disks/RAID on these servers either so unsure why we're getting no output.

Grafana.ini (with some obvious details obfuscated) is attached.
[grafana.txt](https://github.com/grafana/grafana/files/1506249/grafana.txt)

**Comment by gfl3** [comment]
I’m seeing grafana v4.6.2 on Centos 7.4 crash occasionally.  We have prometheus data feeds.

**Comment by johnewing1** [comment]
I've seem similar behaviour, and the root cause was bad dns performance.

In our case we used an external postgres db for both sessions and db storage, and grafana would completely hang with no log output and and no cpu usage.

We eventually diagnosed this by using strace to find that it was hanging trying to connect to the db.  When the db was obviously ok, we then checked other parts of the network stack and found dns resolution would occasionally take seconds and this would hang grafana.

**Comment by ckorobov** [comment]
Same here (grafana 4.6.2, was upgraded 10 days ago), on Centos 6.7, no network dependencies (we have sqlite and redis for sessions on the same host). Also it's not hanging, it's being crashed (no process running), and nothing in logs. We are using upstart to restart it as a workaround, and thinking of downgrading to earlier minor version.

**Comment by ckrybus** [comment]
I have the same problem, 3 hours ago I did a grafana upgrade from 4.0.2 to 4.6.3 and since then grafana crashed two times already. There is no stack trace in the log.


Datasource: Graphite
DB: Sqlite
OS: CentOS release 6.8 (Final)

Upgrade info:
```
---> Package grafana.x86_64 0:4.0.2-1481203731 will be updated
---> Package grafana.x86_64 0:4.6.3-1 will be an update
```

After both crashes the last line in the log was:

```
t=2017-12-19T15:45:52+0100 lvl=info msg="Database table locked, sleeping then retrying" logger=sqlstore retry=0
```

but I am not sure if it has something to do with it.

**Comment by ckorobov** [comment]
I believe it is related:
https://github.com/mattn/go-sqlite3/issues/51 and https://github.com/jmoiron/sqlx/issues/120
We've set this:
GOMAXPROCS=1 
and so far there are no crashing.

**Comment by pasancario** [comment]
We're having same behaviour with Grafana 4.61 with SQLITE DB but running on a Debian 7. 

Did GOMAXPROCS need to be changed in compile time? Cannot be modified directly in Grafana App right?

**Comment by bedrin** [comment]
Looks like `ENV GOMAXPROCS 1` did the trick.
I don't see crashed any more, although not 100% sure.

@pasancario just set this environment variable before launching Grafana server. It is **NOT** a compile time setting.

**Comment by Arovix** [comment]
I was just checking in on the linked ticket (#9706) and wanted to clarify that I wasn't seeing anything printed to the log files either when it crashed. After it crashed a couple times I started the server as a foreground process and left the shell open, which is where the stack trace would print once it crashed.

I never did get it working, but I'll try the GOMAXPROCS thing and report back.

**Comment by Arovix** [comment]
Setting GOMAXPROCS=1 before starting the serve worked for me too. It's been two weeks now on 4.6.3 with no crashes.

**Comment by phateuk** [comment]
@Arovix - Did you set this in the init file for grafana-server? Our init file has a bunch of ENVs near the top of the file i.e:

GRAFANA_USER=grafana

and so on.. can GOMAXPROCS=1 be set at the same point or does it need to be elsewhere?

**Comment by Arovix** [comment]
@phateuk I'm not sure about the init file, we don't set any other environment variables that way so I didn't try it. I just set it in the terminal before running the command to start the server. 

So for bash it looked like:
```
From the location of your grafana install
$ sudo su
$ export GOMAXPROCS=1
$ bin/grafana_server -config conf/config.ini &
```

**Comment by torkelo** [comment]
I think this has been fixed, no recent reports of crashes in v5.0+
